src/pymmcore_plus/mda/_engine.py

```diff
@@ -567,6 +580,42 @@ def _set_event_z(self, event: MDAEvent) -> None:
         correction = self._z_correction.setdefault(p_idx, 0.0)
         self._mmc.setZPosition(cast("float", event.z_pos) + correction)
 
+    def _set_event_slm_image(self, event: MDAEvent) -> None:
+        try:
+            # Get the SLM device
+            if not (slm_device := event.slm_image.device or self._mmc.getSLMDevice()):
+                raise ValueError("No SLM device found or specified.")
+
+            # cast to numpy array
+            slm_array = np.asarray(event.slm_image)
+            # if it's a single value, we can just set all pixels to that value
+            if slm_array.ndim == 0:
+                value = slm_array.item()
+                if isinstance(value, bool):
+                    on_value = _SLM_DEVICES_PIXEL_ON_VALUES.get(slm_device, 1)
+                    value = on_value if value else 0
+                self._mmc.setSLMPixelsTo(slm_device, int(value))
+            elif slm_array.size == 3:
+                # if it's a 3-valued array, we assume it's RGB
+                r, g, b = slm_array.astype(int)
+                self._mmc.setSLMPixelsTo(slm_device, r, g, b)
+            elif slm_array.ndim in (2, 3):
+                # if it's a 2D/3D array, we assume it's an image
+                # where 3D is RGB with shape (h, w, 3)
+                if slm_array.ndim == 3:
+                    if not slm_array.shape[2] == 3:
+                        raise ValueError(
+                            "SLM image must be 2D or 3D with 3 channels (RGB)."
+                        )
+                # convert boolean on/off values to pixel values
+                if slm_array.dtype == bool:
+                    on_value = _SLM_DEVICES_PIXEL_ON_VALUES.get(slm_device, 1)
+                    slm_array = np.where(slm_array, on_value, 0).astype(np.uint8)
+                self._mmc.setSLMImage(slm_device, slm_array)
```

Comment:
I think we *may* want to call `self._mmc. displaySLMImage(slm_device)` here:

cc @marktsuchida, I'm not entirely sure when/whether one is supposed to call `displaySLMImage`.  Should we be doing it here immediately after setting it?  I looked through both the clojure and AcqEng and didn't see any calls to it.  The only thing I found was [here in SLM.java](https://github.com/micro-manager/micro-manager/blob/2244df2cc6e29e12221ff8c86702b9d32dc808d7/libraries/Projector/src/main/java/org/micromanager/projector/internal/devices/SLM.java#L167-L177), where pretty much every call to `setSLMImage` is followed immediately by `displaySLMImage`

Reply:
when I run this all locally, using a computer monitor as an SLM, I definitely don't see anything when running the example code that @hinderling provided in the post above unless i manually call `displaySLMImage` here.  seems fine to do it here, but we could also delay that to `exec_event` as well if there's any reason to wait

Reply:
I've added it to `exec_single_event` and `exec_sequenced_event`. If there is an exposure time set, the countdown would otherwise already start in the `setup_event`, my hope is to minimize camera/dmd exposure offset. Could you check if on your setup the image gets displayed now, even for sequenced acquisitions? 
Here I'm not sure what to do: Do we need to re-trigger `DisplaySLMImage` for every frame in a sequenced acquisition? 

Called when executing a single/sequenced event:
```python
def _exec_event_slm_image(self, event: MDAEvent) -> None:
    if not event.slm_image:
        return
    if slm_device := event.slm_image.device or self._mmc.getSLMDevice():
        try:
            self._mmc.displaySLMImage(slm_device)
        except Exception as e:
            logger.warning("Failed to set SLM Image: %s", e)
```

If two slm_images are the same, the events are sequenceable. Current check:
```python
if e1.slm_image != e2.slm_image:
    return _nope("SLMImage sequences are not supported yet.")
```

if `DisplaySLMImage` needs to be re-triggered for every acquisition, we could change it to:
```python
    if e1.slm_image is None or e2.slm_image is None:
        return _nope("SLMImage sequences are not supported yet.")
```
This would break however any sequencing with an SLM image, even for DMDs that don't need re-triggering. 
With the current implementation for me displaying `slm_img_0, slm_img_0, slm_img_0, slm_img_0` is faster than `slm_img_0 ,slm_img_1 ,slm_img_0, slm_img_1`, as the first sequence can be sequenced.

I think my confusion is stemming from not understanding sequencing. If we acquire a burst of 10 images, in a sequence, how many times is `exec_sequenced_event` executed?

Reply:
Seems to work correctly with genericSLM device now! Still not clear what happens to SLMs where SLMExposure actually has an effect. Images are displayed forever on Polygon1000 and genericSLM, no matter the SLMExposure.